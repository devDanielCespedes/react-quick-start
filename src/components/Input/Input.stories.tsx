import type { Meta, StoryObj } from "@storybook/react";
import { Input } from "./Input";
import { TextFieldProps } from "@mui/material";
import { fn } from "@storybook/test";
import { withActions } from "@storybook/addon-actions/decorator";

const meta: Meta<typeof Input> = {
  title: "Components/Input",
  component: Input,
  tags: ["autodocs"],
  parameters: {
    layout: "centered",
    actions: { argTypesRegex: "^on.*", handles: ["focus", "blur", "change"] },
  },
  argTypes: {
    label: { control: "text" },
    variant: { control: "select", options: ["outlined", "filled", "standard"] },
    error: { control: "boolean" },
    helperText: { control: "text" },
    disabled: { control: "boolean" },
    size: { control: "select", options: ["small", "medium"] },
    fullWidth: { control: "boolean" },
    onChange: fn(),
    onFocus: fn(),
    onBlur: fn(),
  },
  decorators: [withActions],
};

export default meta;
type Story = StoryObj<TextFieldProps>;

export const Outlined: Story = {
  args: {
    label: "Outlined TextField",
    variant: "outlined",
  },
};

export const Filled: Story = {
  args: {
    label: "Filled TextField",
    variant: "filled",
  },
};

export const Standard: Story = {
  args: {
    label: "Standard TextField",
    variant: "standard",
  },
};

export const WithError: Story = {
  args: {
    label: "With Error",
    variant: "outlined",
    error: true,
    helperText: "Error message",
  },
};

export const Disabled: Story = {
  args: {
    label: "Disabled TextField",
    variant: "outlined",
    disabled: true,
  },
};

export const Small: Story = {
  args: {
    label: "Small TextField",
    variant: "outlined",
    size: "small",
  },
};

export const Medium: Story = {
  args: {
    label: "Medium TextField",
    variant: "outlined",
    size: "medium",
  },
};

export const FullWidth: Story = {
  args: {
    label: "Full Width TextField",
    variant: "outlined",
    fullWidth: true,
  },
};
