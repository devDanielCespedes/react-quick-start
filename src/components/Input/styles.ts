import { SxProps } from "@mui/system";

export const styles: SxProps = {
  width: {
    xs: "90vw",
    sm: "60vw",
    md: "40vw",
    lg: "30vw",
    xl: "20vw",
  },
  height: "70px",
};
