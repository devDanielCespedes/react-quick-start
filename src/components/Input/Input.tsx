import React from "react";
import { TextFieldProps, TextField } from "@mui/material";
import { styles } from "./styles";

const Input = React.forwardRef<HTMLInputElement, TextFieldProps>(
  (props, ref) => {
    return <TextField sx={styles} {...props} inputRef={ref} />;
  }
);

Input.displayName = "Input";

export { Input };
