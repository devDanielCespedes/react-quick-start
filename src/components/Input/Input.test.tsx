import { render, screen, fireEvent } from "@testing-library/react";
import { Input } from "./Input";

test("renders Input with label", () => {
  const { asFragment } = render(<Input label="Test Label" />);
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  expect(textFieldElement).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

test("Input onChange works", () => {
  const handleChange = jest.fn();
  const { asFragment } = render(
    <Input label="Test Label" onChange={handleChange} />
  );
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  fireEvent.change(textFieldElement, { target: { value: "new value" } });
  expect(handleChange).toHaveBeenCalledTimes(1);
  expect(asFragment()).toMatchSnapshot();
});

test("Input onFocus works", () => {
  const handleFocus = jest.fn();
  const { asFragment } = render(
    <Input label="Test Label" onFocus={handleFocus} />
  );
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  fireEvent.focus(textFieldElement);
  expect(handleFocus).toHaveBeenCalledTimes(1);
  expect(asFragment()).toMatchSnapshot();
});

test("Input onBlur works", () => {
  const handleBlur = jest.fn();
  const { asFragment } = render(
    <Input label="Test Label" onBlur={handleBlur} />
  );
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  fireEvent.blur(textFieldElement);
  expect(handleBlur).toHaveBeenCalledTimes(1);
  expect(asFragment()).toMatchSnapshot();
});

test("Input with error message", () => {
  const { asFragment } = render(
    <Input label="Test Label" error={true} helperText="Error message" />
  );
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  expect(textFieldElement).toBeInTheDocument();
  const helperTextElement = screen.getByText(/Error message/i);
  expect(helperTextElement).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

test("Input disabled", () => {
  const { asFragment } = render(<Input label="Test Label" disabled={true} />);
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  expect(textFieldElement).toBeDisabled();
  expect(asFragment()).toMatchSnapshot();
});

test("Input fullWidth", () => {
  const { asFragment } = render(<Input label="Test Label" fullWidth={true} />);
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  expect(textFieldElement).toHaveStyle("width: 100%");
  expect(asFragment()).toMatchSnapshot();
});

test("Input small size", () => {
  const { asFragment } = render(<Input label="Test Label" size="small" />);
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  expect(textFieldElement).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

test("Input medium size", () => {
  const { asFragment } = render(<Input label="Test Label" size="medium" />);
  const textFieldElement = screen.getByLabelText(/Test Label/i);
  expect(textFieldElement).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});
