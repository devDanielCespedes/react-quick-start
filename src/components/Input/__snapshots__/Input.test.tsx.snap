// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input disabled 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r5:"
      id=":r5:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled=""
        id=":r5:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input fullWidth 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-17ajlrd-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r6:"
      id=":r6:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r6:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input medium size 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r8:"
      id=":r8:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r8:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input onBlur works 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r3:"
      id=":r3:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r3:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input onChange works 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r1:"
      id=":r1:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r1:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input onFocus works 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r2:"
      id=":r2:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r2:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input small size 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r7:"
      id=":r7:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r7:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input with error message 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r4:"
      id=":r4:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-describedby=":r4:-helper-text"
        aria-invalid="true"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r4:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
      id=":r4:-helper-text"
    >
      Error message
    </p>
  </div>
</DocumentFragment>
`;

exports[`renders Input with label 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1dlczlc-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r0:"
      id=":r0:-label"
    >
      Test Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Test Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
